modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts

```diff
@@ -79,13 +94,70 @@ const viewportUpdate = WindowVisualViewport.get().fold(
   }
 );
 
-const toggleFullscreen = (editor: Editor, fullscreenState: Cell<any>) => {
+const getFullscreenRoot = (editor: Editor): SugarElement<Element> => {
+  const elem = SugarElement.fromDom(editor.getElement());
+  return SugarShadowDom.getShadowRoot(elem).map(SugarShadowDom.getShadowHost).
+    getOrThunk(() => SugarBody.getBody(Traverse.owner(elem)));
+};
+
+const getFullscreenElement = (root: DocumentOrShadowRoot) => {
+  if (root.fullscreenElement !== undefined) {
```

Comment:
Would it be worth defining types for these, rather than using any? Something like this:

```ts
interface NativeFullscreen {
  fullscreenElement: Element;
  requestFullscreen: () => void;
  exitFullscreen: () => void;
}

interface MsFullscreen {
  msFullscreenElement: Element;
  msRequestFullscreen: () => void;
  ...
}

...

type Fullscreen = NativeFullscreen | MsFullscreen | ...

const isNative = (fs: Fullscreen): fs is NativeFullscreen = Obj.has(fs, 'fullscreenElement');
...
```

It's a bit more verbose, but you'd be able to replace all of the `(root as any)` instances with a single `(root as any as Fullscreen)`, after which point typescript would be able to track all of the types throughout the plugin.

Short of doing that, could we add explicit return types to these functions like `getFullscreenElement`, as otherwise the `any` will propagate through things.

Reply:
You're really not getting anything from doing that though... it saves one check. If you were to do this, you'd be better casting the element instead. eg:

```
interface FullscreenElement extends Element. {
  msFullscreenElement?: Element;
  msRequestFullscreen?: () => void;
  webkitFullscreenElement?: Element;
  webkitRequestFullScreen?: () => void;
}
```

Reply:
Worth noting though historically we've gone with the `any` approach for browser specific prefixes like this.

Reply:
You can also override the interface directly to add items (we sometimes do that for Window for globals we add). So it would become:
```
interface DocumentOrShadowRoot {
msFullscreenElement?: Element;
webkitFullscreenElement?: Element;
msExitFullscreen?: () => void;
webkitCancelFullScreen?: () => void;
}
interface Element {
msRequestFullscreen?: () => void;
webkitRequestFullScreen?: () => void;
}
```
https://www.typescriptlang.org/docs/handbook/declaration-merging.html

Reply:
Hmm, doesn't seem to work. Nevermind.

Reply:
I have annotated the return types on all the functions.

Reply:
I had another go at the declaration merging and this works:
```
declare global {
  interface DocumentOrShadowRoot {
    msFullscreenElement?: Element | null;
    webkitFullscreenElement?: Element | null;
  }
  interface Document {
    msExitFullscreen?: () => void;
    webkitCancelFullScreen?: () => void;
  }
  interface Element {
    msRequestFullscreen?: () => void;
    webkitRequestFullScreen?: () => void;
  }
}
```
Note that the change in the types seems to extend outside the file (which I didn't expect). So we may not want this. Should I leave it as is?

Reply:
Yeah, I thought that's what you'd have to do to get declaration merging working. I'd vote to not do that, because as you said it changes the global types, not just in this file.